멤버 주간 기록 api 작업

Na2te opened the pull request:
## 📌 연결된 이슈
- #66 

### ✅ 작업 내용
- 멤버의 주간 공부 기록 슬라이싱 작업
- 멤버의 태그별 누적 공부량 api 연동

### 📷 관련 이미지(영상)

|<img width="300" alt="image" src="https://github.com/user-attachments/assets/09172002-a2e3-4136-b635-a14fb652f8ac" />|<img width="300" alt="image" src="https://github.com/user-attachments/assets/c6d01af1-b7f4-444b-b5dd-aa730df506de" />|

![image](https://github.com/user-attachments/assets/0cae2382-8689-49fb-afbe-13f2a926ad2e)

### 🤔 의논하고 싶은 내용(코드 리뷰 받고 싶은 부분)
- 처음 호출 시에는 꼭 에러가 뜨던데 조금 더 이유를 찾아보겠습니다
- 1번째 사진처럼 에러가 났을 때 Text 부분 누르면 다시 갱신합니다
- 현재는 데이터가 없어서 아무것도 없을 때는 2번째 사진처럼 나오도록 해놨습니다
- 무아지경으로 코드를 짜버려서 멍멍이 같이 짠 거 같은데 의견 주시면 감사하겠습니다

Na2te committed:
fix(#56): StudyRecordList 그림자 수정
- SpaceBy로 되어 있어 마지막 아이템은 그림자가 잘리는 문제 수정

Na2te committed:
feat(#56): Record 스크린 UI 구햔
- 카드까지 적용 완료

Na2te committed:
fix(#66): 로그인 에러 수정
- UserDetailResponse의 nickname이 camel case가 아니라 발생하는 에러 수정

Na2te committed:
fix(#66): 안쓰는 것 삭제
- 더 이상 안쓰는 RecordTopAppBar 삭제

Na2te committed:
feat(#66): api 추가
- 멤버의 주간 태그별 공부량 및 주간 공부 컨텐츠 가져오는 api 함수 작성 및 관련 데이터 클래스 작성

Na2te committed:
docs(#66): StudyGroupsApi 주석 수정
- 학습 내용 반환하는 함수 주석을 수정 안해서 원래 기능에 대한 설명으로 수정

Na2te committed:
feat(#66): StudyGroup Api 의존성 추가
- 새로 만든 StudyGroup Api 의존성 생성

Na2te committed:
fix(#66): Api 모델 위치 변경
- 태그별 공부 시간 등 관련 API에 필요한 model을 studyies에서 studygroup으로 변경

Na2te committed:
feat(#66): APi 작업 완료
- StudyGroup DataSource 및 API Return Typd 잘못 설정한 것 수정
- StudyGroupDataSource 의존성 bind 추가
- DataSourceImpl 구현

Na2te committed:
feat(#66): paging3 라이브러리 추가

Na2te committed:
style(#66): 도메일 모델 위치 수정
- studygroup의 모델 위치가 잘못된 것 수정

Na2te committed:
feat(#66): 스터디 그룹 멤버 주간 공부기록 API 작업 완료
- Datasource, Repository, Module, UseCase 등 추가 및 타입 수정

Na2te committed:
feat(#66): 타임 포맷터 함수 추가
- api로 Long값의 태그별 학습시간을 포맷에 맞춰 String 값으로 변화시키는 함수 작성

Na2te committed:
fix(#66): PageConfig 변경
- placeholder를 false로 변경

Na2te committed:
feat(#66): 멤버 공부 화면 API 연동
- 기존 MemberId와 GroupId가 Int이던 것 Long으로 수정
- Paging3의  flow 데이터를 State로 넣는 건 좀 아닌 거 같아서 별도 변수로 작성

Na2te committed:
fix(#66): 모델 추가 및 수정
- 멤버 기록 보는 것 관련 수신 객체를 data 모델과 domain 모델로 분리

Na2te committed:
fix(#66): API 데이터 타입 수정
- api의 list가 persistentList로 매핑이 안되어서 List로 수정

Aloe-droid commented on app/src/main/java/com/ssafy/neegongnaegong/data/paging/MemberStudyContentsPagingSource.kt:
```diff
@@ -0,0 +1,43 @@
+    }
+
+    override fun getRefreshKey(state: PagingState<MemberStudyContentSliceKey, StudyContentInfo>): MemberStudyContentSliceKey? {
+        return state.anchorPosition?.let { state.closestPageToPosition(it)?.nextKey }
```
refreshKey는 현재 키를 가져오는게 목적인데, 항상 nextKey를 return 하게 되면 다음 키를 전달할 수도 있지 않을까여 
```
state.anchorPosition?.let { anchorPosition ->
      val anchorPage = state.closestPageToPosition(anchorPosition)
      anchorPage?.prevKey?.plus(1) ?: anchorPage?.nextKey?.minus(1)
    }
```
위 코드는 구글 예제 코드입니당

Na2te replied:
맞는 말입니다
우선 저 예제는 오프셋 기반 페이지네이션을 구현한 것이기에, 현재 위치의 키를 next에서 1을 빼거나, prev에서 1을 더하는 식으로 구하는 것인데
저희는 커서 기반 페이지네이션이기 때문에, next와 prev 데이터 간 연관관계가 없어지는 거로 알고 있습니다. (특히 저희가 쓰는 커서 키는 날짜와 기록 id를 Key로 사용합니다)
그래서 현재 키가 정확하게 무엇인지 접근할 방법이 없고, 현재 API상 prev key는 제공하지 않습니다
그래서 그냥 key를 null로 주거나 nextKey를 넘겨주는 것 말고는 떠오르는 방안이 없었습니다

refreshKey가 꼭 현재 키를 줘야한다기보다, 현재 ViewPort내에 있는 값들이 다시 갱신될 수 있게 하는 Key면 되는 것으로 알고 있어서 조금 더 방법을 찾아보겠습니다

Na2te replied:
Vibe 코딩에 너무 생각 없이 있었던 것 같습니다
state에서 아이템에 접근이 가능하네요
여기에 있는 데이터로 Key 구성했습니다
반성하게 되네요
이건 어떤가요??
수정한 결과입니다
현재 Paging에서 가지고 있던 데이터(공부기록 리스트)의 공부기록마다 러닝 레코드 id하고 시간이 있는데 커서도 이것 중에서 마지막 값 다음에 해당하는 데이터의 키를 준다던지 하는 식으로 nextKey 값을 던져주는게 아닐까 싶어서 이거 제일 처음 아이템 기준으로 해봤습니다
근데 문현이 커서가 create 기준인지 modified 기준인지 모르겠어서 물어봐야겠네요
```kotlin
override fun getRefreshKey(state: PagingState<MemberStudyContentSliceKey, StudyContentInfo>): MemberStudyContentSliceKey? {
        return state.firstItemOrNull()?.let {
            MemberStudyContentSliceKey(
                it.learningRecordCreatedAt,
                it.learningRecordId
            )
        }
    }
```

Aloe-droid replied:
너무 좋은것 같슴다 ㅎㅎ

Na2te replied:
날카로운 지적 감사합니당 :)

limnyn replied:
@Na2te

지금은 
1~n까지 반환 시 공부목록리스트와
n번째 객체의 db상 cursorId 및 cursorCreadtedAt을 반환합니다

`cursorCreatedAt` : **그룹의 피드 최신순 정렬**을 위한 **컬럼**  != 공부기록 생성시점

따라서 이를 구분하기 위해 공부기록 생성/변경시점은 learningRecordCreatedAt 및 learningRecordModifedAt 을 썼습니다

요약하자면
1. 공부기록생성시간과 그룹별 피드 생성시간은 다르다(값도 다르고 사용처도 다르다)
2. 공부기록 생성시간 `learningRecordCreatedAt` : 공부기록 생성된 시점(공부종료시점)
3. `cursorCreatedAt` : 공부기록 **생성 후** 그룹게시판에 연결된 시점. 따라서 ms 차이 발생. 공부기록과 상관없이 피드출력을 위해 씀


현재는 마지막 피드의 cursorId, cursorCreatedAt을 반환하는데
만약 목록 중간의 특정 피드에 대한 로직이 필요하시다면 백엔드에서는 두가지 방법으로 가능합니다
1. 백엔드에서 모든 피드에 대해 cursorId, cursorCreatedAt을 반환해주도록 responseDto 수정
2. 재호출할 피드의 learningRecordId를 전달하면 해당 피드 값만 반환해주는 api 추가 (현재 개인공부기록 상세조회는 작성자만 조회 가능)

Na2te replied:
@limnyn 흠 1번으로 혹시 가능할까요..?

limnyn replied:
적용 및 테스트 후에 pr에 멘션드리겠습니다

Na2te replied:
감사합니다!

limnyn replied:
@Na2te 
작업 및 테스트를 해봤는데 이게 굳이 필요할까 싶어서 로컬에만 들고 있고 아직 pr은 안올렸습니다. 

1. api 반환에 불필요한 반복이 많이 일어나고 
2. 특히 피드-공부기록 쪽 제외하고는 기존필드와 중복되는 경우가 많습니다.
ex) 스터디그룹 목록의 경우 기존 필드인 createdDate + studyGroupId 값이 사실상 cursor값입니다

paging3를 잘 모르는데 append할 때 필요한 커서값은 이미 포함되있고 refresh도 저희는 수동 새로고침밖에없는걸로 알고 있어서 null값으로 커서값 넣으면 되고, 중간값 커서 정보는 쓰는 경우가 없지않나요?

제가 이부분 잘 모르는데 혹시 이 부분 안드분들끼리 실제로 쓰는값인지 또는 기존 필드들로 대체 가능한지, 언제쓰이는지 의논 후 설명 가능한가요?

만약 필요한 기능이라면 합의해서 공통된 규격으로 해당값 전달하도록 필드 통일해서 적용해보겠습니다.
-> 커서용 필드 추가 대신 목록아이템들의 id, createdAt을 땡겨와서 커서 호출 시 사용할 수 있도록 백엔드 응답을 통일할 것 같습니다.

커서쪽 필드 이름을 통일해야 안드에서 재사용하기 편할 것 같습니다.

Na2te replied:
@limnyn 
Paging3 자체에서 refreshKey를 요구하기 때문이긴 합니다
요구하는 이유로는 함수에 적혀져 있는 설명 및 글을 보자면 
![image](https://github.com/user-attachments/assets/db5aae57-2f8b-45a2-afe8-9c471265eb91)
이러합니다
[Paging3 소개 링크](https://medium.com/androiddevelopers/introduction-to-paging-3-0-in-the-mad-skills-series-648f77231121)

결론적으로 데이터베이스 업데이트든, 프로세스가 죽었든(홈 화면 같은 다른 거 하다가 앱이 리프레시 일어나거나 했을 때 말하는 거겠죠..? 틀렸다면 지적해주세용), 화면 회전과 같은 구성 변화든,  새로고침이든 기타의 이유로 데이터를 새로 불러오기 위함입니다. 그리고 일반적으로 현재 보고 있던 부분의 데이터를 중심으로 새롭게 로딩을 하고자 합니다. (화면 변경 같은 거 일어날 때마다 계속 처음부터 다시 로딩할 게 아니기에)

인덱스 기반 페이지네이션이면 그냥 next -1 같은 값을 통해 현재 보던 곳의 내용이 포함된 값으로 갱신을 해서 위 목적을 충족할 수 있는데, 저희는 커서 기반이라 next, prev의 키 값이 현재 키랑 관련이 없어서 알 수 있는 방법이 없습니다.

아니면 차선책으로 nextKey를 API에서 건네준 cursorCreatedAt, cursorId로 하고, prev를 그냥 현재 키 값으로 넣어둔 다음에, refreshKey로 prev값(잘 들어갔으면 현재 키 값일 거고, 만약 안들어가면 null)으로 그냥 할까요?

만약 화면 회전 등과 같은 변경 시에도 그냥 처음부터 다시 보는 것이 의도대로라면, 그냥 refreshKey를 무조건 null로 줘서 갱신을 해야 하는 상황이 생기면 무조건 새로고침 하도록 하는 것이 맞긴 합니다

Na2te committed:
fix(#66): PagingSource RefreshKey 수정
- 현재 Paging의 Key로 수정

todayis-sunny commented on app/src/main/java/com/ssafy/neegongnaegong/data/repository/StudyGroupRepositoryImpl.kt:
```diff
@@ -0,0 +1,37 @@
+            }
+        ).flow
+
+}
```
`domain`모듈에서 android의존성을 없애기 위해서 `data`모듈에서 **paging**을 활용하시게 된걸까요?

Na2te replied:
안드 의존성이 없는 paging3 라이브러리가 있긴한데, 저희가 레이어 기반 멀티모듈를 적용한게 아니라서 Paging3의 어느 부분까지가 안드와 엮여있는건지 확신이 없어 최대한 데이터 모듈에 밀어넣은거긴 합니다

코드적으로도 뭔가 domain에서 처리할 내용같아보이지는 않아가지구용

사실 코드 작성할 때 그렇게 큰 고려를 하진 않았습니다

todayis-sunny commented on app/src/main/java/com/ssafy/neegongnaegong/presentation/group/record/component/StudyRecordList.kt:
```diff
@@ -0,0 +1,109 @@
+            }
+        }
+
+        when {
+            lazyItems.loadState.refresh is LoadState.Loading -> item { LoadingItem() }
+            lazyItems.loadState.append is LoadState.Loading -> item { LoadingItem() }
+            lazyItems.loadState.refresh is LoadState.Error -> item {
+                val e = lazyItems.loadState.refresh as LoadState.Error
+                ErrorItem(e.error.localizedMessage.orEmpty()) { lazyItems.retry() }
+            }
+            lazyItems.loadState.append is LoadState.Error -> item {
+                val e = lazyItems.loadState.append as LoadState.Error
+                ErrorItem(e.error.localizedMessage.orEmpty()) { lazyItems.retry() }
+            }
+            lazyItems.itemCount == 0 -> item {
+                NoDataItem()
+            }
```
`LazyPagingItems`이 **paging**과 짝궁의 느낌인거죠?
`loadState`를 통해 분기처리 하는게 유용해 보입니다:)

Na2te replied:
Paging데이터는 `collectasatate` 같은 것처럼 collect~~ 함수로 불러오더라구요
Compose에서 Paging 데이터 구독하면 저렇게 되더라구요
분기처리는 gpt가 잘 정리해주더라구요 저도 배웠습니다

todayis-sunny commented on app/src/main/java/com/ssafy/neegongnaegong/presentation/util/TimeFormatter.kt:
```diff
@@ -90,4 +91,19 @@ object TimeFormatter {
         return time + offsetMillis
     }
+
+    /**
+     * 밀리초로 들어온 기간을 00H00M으로 바꾸기 위한 함수
+     * @param Long 기간
+     * @return String 00H00M 꼴
+      */
+    @SuppressLint("DefaultLocale")
+    fun formatDurationToHM(millis: Long): String {
+        val totalMinutes = millis / (1000 * 60)
+        val hours = totalMinutes / 60
+        val minutes = totalMinutes % 60
+
+        return String.format("%02dH%02dM", hours, minutes)
+    }
```
저도 이게 필요한데 `feat/#66-record`에서 해당 커밋을 **체리픽** 활용하는게 좋을까요?

Na2te replied:
깃 고수시네요
Develop에 머지된 이후 작업하시는게 아니라면 말씀하신게 맞습니다!

todayis-sunny reviewed (approved):
작업 고생하셨습니다:)

upsk1 commented on app/src/main/java/com/ssafy/neegongnaegong/domain/usecase/studygroup/GetMemberStudyContentsUseCase.kt:
```diff
@@ -0,0 +1,14 @@
+class GetMemberStudyContentsUseCase(
+    private val repository: StudyGroupRepository,
+) {
+    operator fun invoke(request: StudyMemberInfo): Flow<PagingData<StudyContentInfo>> =
+        repository.getMemberStudyContents(request)
+}
```
저희앱이 전체적으로 UseCase가 invoke만 포함하고 있는데 혹시 영민님은 어떤 비즈니스 로직을 추후에 넣을지 생각해보신게 있으실까요 ?

Na2te replied:
음 글쎄요
`invoke`만 포함하고 있는게 그렇게 문제는 아니라고 생각합니다
비즈니스 로직을 작성하면서 가독성을 위해서 `invoke` 내에 실행되는 어떤 로직을 함수로 뺄 수는 있긴 하겠지만, `Usecase`를 실행하는 건 `invoke()` 함수만이어야 `Usecase`의 정의에 맞지 않을까요??

다만 `invoke`의 코드 자체만 놓고 보자면 확실히 지금 코드는 그저 `repository`의 함수를 그냥 실행하는 정도에 불과하긴 합니다
`repository`의 함수를 실행하는 정도에 끝나는 게 아쉽다? 같은 느낌으로 말씀하신 거라면, `Usecase`마다 달라질 수 있긴 한 거 같습니다

![image](https://github.com/user-attachments/assets/0b17e8e9-6f00-4350-9a62-abe74c8c03d5)

궁금해서 물어봤더니 이런 식으로 하나의 `Usecase`를 위해 여러 `Repository`를 사용하는 경우라던지 등의 경우에는 별도의 코드가 조금 들어갈 수 있긴 하더라구요

그래서 답변은 제가 뭘 더 넣겠다 해서 넣을 수 있는 건 아닌 것 같고, `Usecase`의 필요에 따라서 자연스레 추가되지 않을까..? 하는 생각이었습니다!

upsk1 replied:
맞아요 사실 

> invoke만 포함하고 있는게 그렇게 문제는 아니라고 생각합니다

어떤 관점에서 말씀해주신건진 모르겠지만, Usecase에 저희 앱 비즈니스로직을 녹일 수 있음에도 불구하고 그저 invoke로만 썻다면 사실 큰 문제라고 생각하고있거든요, 지인 면접에서도 이 부분에 어택을 맞기도 했고 
https://www.youtube.com/watch?v=3mR8_vT7m1U
드로이드 나이츠에서도 한번 다뤘던 주제기도해서요 

그리고 이부분에 대해서 고민이 많은데 추후 로직을 넣을 수 있다면 꼭 Usecase를 활용해보고 싶은게 제 의견입니다

Na2te replied:
자료 되게 감사합니다 여러 생각이 들게 하는 내용이었네요

저 부분에 대해서 말씀하시는 거라면 제 코드도 개선할 부분이 있네요
제 코드를 기준으로 본다면 `ViewModel`에서 nav인자를 널체크를 통해 Null이 아닐 경우에만 넣고 있는데, 이걸 그러면 도메인 모델에서 nullable하게 받고, null인지를 체크할 함수 책임(함수)를 작성하고, 특정 `Usecase`에서 `Nullable`하면 안됐을 때, `repository`의 함수를 실행시키는게 아니라 에러를 `Throw` 하도록 해야 한다던지 할 수 있을 거 같네요

다만 조금 걱정되는 건, 에러 핸들링이 현재 `safeCollect()`를 했을 때 기준으로 몰려 있는 것으로 기억하는데, 현재 제 Paging3를 이용한 Slicing의 경우 `.safeCollect` 같은 호출 없이 사용하게 되어서, 이 경우 에러를 터트리면 공통 에러 핸들링 로직을 거쳐가지 않네요. 그냥 `try-catch`, `runCatching` 같은 것을 예외적으로 적용하고 치워야 할 지 어떡할지 고민이 됩니다.

`invoke`로만 썼다는게 아직 무슨 말씀인지는 정확하게 모르겠는데, 
`Usecase.A()`, `Usecase.B()` 이런 식으로 쓸 수도 있어야 한다 같은 말씀이라면 이건 아니라고 생각합니다.
다만 `repository`의 함수 실행을 그저 감싸는 용도 정도로 쓰는 것에 대해 경각심을 의미하는 거였다면, 백 번 공감합니다. 

양질의 좋은 자료 감사합니다
코드 좀 바꿔볼 생각이 강하게 드네요

upsk1 replied:
`invoke` 에 대해서 말씀드린건 백번 공감하시는 경각심을 가져보자 라는 의미가 맞습니다! 

paging3 관련해서는 우선 영민님 의견이 되게 좋아 보입니다 일단 많은 부분에서 paging3가 쓰이고 있지 않아서 예외 적용을 하되 같이 어떻게 효율적으로 처리할 수 있을지 고민해 나가는것도 좋아보입니다!

Na2te committed:
fix(#66): JWT Reissue 수정
- token 저장 시 특정 함수에서는 removePrefix를 하고, 특정 함수에서는 안해서 token의 bearer 값이 제대로 저장되지 않던 문제 수정

Na2te committed:
fix(#66): API 주소 수정
- Path 및 Query를 camel case에서 kebab case로 변경
- cursor Query의 이름 수정

Na2te committed:
fix(#66): domain model 수정
- MemberStudyContentsInfo의 cursor key 관련 속성의 이름을 변경된 것으로 수정
- StudyContentInfo에 cursor key 관련 속성 추가 및 형변환 코드를 멤버함수에서 확장함수로 변경

Na2te committed:
fix(#66): data model 수정
- StudyContentResponse에 cursor key 추가 및 형변환 코드 확장함수로 변경
- StudyLogByTagResponse의 형변환 코드 확장함수로 변경

Na2te committed:
fix(#66): datasource, repository 수정
- Model의 속성 수정으로 NetworkStudyGroupDataSourceImpl의 request의 cursor키 속성 수정
- StudyGroupRepositoryImpl애서 Model의 형변환 확장함수 import

Na2te committed:
fix(#66): Pagingsource 수정
- Data Model의 cursor 키 수정한 것 적용
- Refresh key 관련 로직 수정 (현재 페이지의 가장 앞에 있는 아이템의 cursor key를 이용해서 보던 부분 다시 refresh 하도록 함)

Na2te committed:
style(#66): Preview 수정
- Model 속성 관련해서 Preview에 필요한 데이터 추가

Na2te committed:
Merge branch 'develop' into feat/#66-record

Na2te commented:
중점적으로 확인하실 건 JWT 관련 수정을 했습니다
JWT가 reissue 처리할 때는 Prefix 처리를 하고 어떤 때는 안하고 그래서, Prefix 관련 처리 로직을 전부 saveToken 내부에 넣어놨습니다
앱 처음 빌드하고 api 호출할 때 에러 생기던 이유가 이것 때문이더라구요
결과적으로 수정해서 이제는 한번에 잘 됩니다

나머지는 @limnyn님이 수정해준대로 api 주소 수정하고, cursor 관련 변경된 값 추가했습니다
그리고 **Refresh Key** 관련해서 현재 지켜보고 있는 Page 단위의 데이터들이 전부 갱신되야 할 것 같아서 **넘겨받은 공부내역 List 중 맨 앞의 아이템이 가지고 있는 Cursor Key를 Refresh Key로 사용**하게 했습니다

Na2te commented:
혹시 별 얘기 없으시면 주말에는 일정 있으시니까 월요일 저녁 즈음에 Merge 하도록 하겠습니다
더 의논하고 싶으신 것 있으시다면 언제든지 환영입니다

Na2te committed:
Merge branch 'develop' into feat/#66-record
